### DISCUSSION TITLE
Erwic: Prevent Navigating Away From a Web App

### DISCUSSION THREAD

**Discussion by iss-sj** [Jelmerro/Vieb#566] [Q&A]
I love using Vieb to handle my communication apps, because it has splits and Erwic. Erwic page suggests you can use Vieb to replace other Electron desktop apps: https://github.com/Jelmerro/Vieb/blob/master/Erwic.md#electron-apps

So, I use Erwic to load all apps that I need, each one in its own split.

What happened to me a few times is:
1. I'm in the middle of a call in e.g. Teams
2. Someone shares a link in a call chat
3. I accidentally open it in the same tab ("f" key, startFollowCurrentTab)

This action moves me away from the app I had a call in, i.e. since I open a new page - I'm dropped out of that call.

Is there a way to prevent this from happening? So I can navigate only within that same app, but never leave it? Usually when this happens, I just go back in history, no harm. But it is an issue if there's an ongoing call. So I would like to block that possibility alltogether and have a complete dedicated Electron app experience.

I did my homework and tried to implement this somehow. I wanted to block the transition to the new page and display a warning (a reminder to open an external link in a new tab / external app).

My initial idea was to just write a greasemonkey script and `preventDefault` on `beforeunload`. So, my script is executing, but preventUnload does not have any effect. Then, after some frustrating time of debugging this, I found the following: https://github.com/Jelmerro/Vieb/blob/master/CHANGELOG.md?plain=1#L591

I'm assuming this is the reason my script is not working?

Thanks in advance for any suggestions/tips.

**Comment by iss-sj** [comment]
To clarify, I'm loading the script as suggested on Erwic page: https://github.com/Jelmerro/Vieb/blob/master/Erwic.md#multiple-pages

Here's the script:

```
const blockLeaving = (event) => {
  event.preventDefault();
  event.returnValue = true;
  // display some alert
};

window.addEventListener("beforeunload", blockLeaving, false);
```

`blockLeaving` is executed, but with no effect. The split transitions to the next page.

**Reply by Jelmerro** [reply]
I can at least confirm that beforeunload calls cannot block page redirects in Vieb, that is true. It's not a unique feature anymore though, as Google plan to remove this API from the browser entirely, Vieb was just a bit earlier and refusing to give this specific power to pages, see here: https://developer.chrome.com/docs/web-platform/deprecating-unload

**Comment by Jelmerro** [comment]
As for your issue, I don't think there's a solution in Vieb for this at the moment, the best we have is the `containernewtab` setting and the related ones (see `:h containernewtab`), but as the name suggests, it's only for new tabs (and split pages + startup pages with the other settings), not for current page navigations. I think this deserves a new issue to add this setting, I'll make one when I'm back to working on Vieb in a few weeks, or feel free to make one in the meantime.
